Annotation formats documentation (#719)
opencv/cvat: cvat/apps/annotation/README.md

  2. match by frame number (if CVAT cannot match by name). If filename tag contains a number CVAT will interpret it as frame number.

- supported shapes: Rectangles
- limitations: Support of Pascal VOC object detection format
- additional comments: the CVAT task should be created with the full label set that may be in the annotation files

#### How to create a task from Pascal VOC dataset
1. Download the Pascal Voc dataset
1. Create the CVAT task with the following labels:
   ```
   aeroplane bicycle bird boat bottle bus car cat chair cow diningtable dog horse motorbike person pottedplant sheep sofa train tvmonitor
   ```
   You can add `~checkbox=difficult:false ~checkbox=truncated:false` attributes if you want to use it.

   Select interesting image files (See [Creating an annotation task](cvat/apps/documentation/user_guide.md#creating-an-annotation-task) guide for details)
1. zip the corresponding annotation files
1. click `Upload annotation` button, choose `Pascal VOC ZIP 1.0` and select the *.zip file with annotations from previous step. It may take some time.

### [YOLO](https://pjreddie.com/darknet/yolo/)

> #### Dumper description

Reviewer: Need an example for YOLO as well. For example, you can redirect the dataset preparation to https://pjreddie.com/darknet/yolo/ article.

  ├── ...
  └──obj.names
  ```
  Each `*.txt` file contains annotations for a frame with the same name as annotation file. Note: it should be possible to match the CVAT frame(imagename) and annotation filename. There are 2 options:
  1. full match between image name and filename form annotation `*.txt` file (in case of a task was created from images or archive of images).
  2. match by frame number (if CVAT cannot match by name). If name of the *.txt annotation file contains a number CVAT will interpret it as frame number.

- supported shapes: Rectangles
- additional comments: the CVAT task should be created with the full label set that may be in the annotation files

### [MS COCO Object Detection](http://cocodataset.org/#format-data)
#### Dumper description
- downloaded file: single unpacked `json`. Detailed description of the MS COCO format can be found [here](http://cocodataset.org/#format-data)
- supported shapes - Polygons, Rectangles (interpreted as polygons)

#### Loader description
- uploaded file: single unpacked `*.json`.
- supported shapes: Polygons
- additional comments: the CVAT task should be created with the full label set that may be in the annotation files

> #### How to create a task from PMS COCO dataset

Reviewer: PMS -> MS. Could you please look at the whole text one more time and fix obvious typos?

1. Download the [MS COCO dataset](http://cocodataset.org/#download). For example [2017 Val images](http://images.cocodataset.org/zips/val2017.zip) and [2017 Train/Val annotations](http://images.cocodataset.org/annotations/annotations_trainval2017.zip).
1. Create a CVAT task with the following labels:
   ```

>    "person" "bicycle" "car" "motorcycle" "airplane" "bus" "train" "truck" "boat" "traffic light" "fire hydrant" "stop sign" "parking meter" "bench" "bird" "cat" "dog" "horse" "sheep" "cow" "elephant" "bear" "zebra" "giraffe" "backpack" "umbrella" "handbag" "tie" "suitcase" "frisbee" "skis" "snowboard" "sports ball" "kite" "baseball bat" "baseball glove" "skateboard" "surfboard" "tennis racket" "bottle" "wine glass" "cup" "fork" "knife" "spoon" "bowl" "banana" "apple" "sandwich" "orange" "broccoli" "carrot" "hot dog" "pizza" "donut" "cake" "chair" "couch" "potted plant" "bed" "dining table" "toilet" "tv" "laptop" "mouse" "remote" "keyboard" "cell phone" "microwave" "oven" "toaster" "sink" "refrigerator" "book" "clock" "vase" "scissors" "teddy bear" "hair drier" "toothbrush"

Reviewer: Do we really need quotes here?

   ```

   Select val2017.zip as data (See [Creating an annotation task](cvat/apps/documentation/user_guide.md#creating-an-annotation-task) guide for details)
1. unpack annotations_trainval2017.zip
1. click `Upload annotation` button, choose `COCO JSON 1.0` and select `instances_val2017.json.json` annotation file. It may take some time.

### [TFRecord](https://www.tensorflow.org/tutorials/load_data/tf_records)

> TFRecord is a very flexible format, but we try to correspond the format that used in [TF object detection](https://github.com/tensorflow/models/tree/master/research/object_detection) with minimal modifications.

Reviewer: Need an example for TFRecord. Use https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_pascal_tf_record.py script to write instructions.

  to reduce version conflicts, etc. Thus, custom code can be run in an extended environment, and core CVAT modules
  should not be affected. As well, this functionality can be useful for Auto Annotation module.

## Format specifications

### CVAT
This is native CVAT annotation format.
[Detailed format description](cvat/apps/documentation/xml_format.md)

#### CVAT XML for images dumper
- downloaded file: Single unpacked XML
- supported shapes - Rectangles, Polygons, Polylines, Points

#### CVAT XML for videos dumper
- downloaded file: Single unpacked XML
- supported shapes - Rectangles, Polygons, Polylines, Points

#### CVAT XML Loader
- uploaded file: Single unpacked XML
- supported shapes - Rectangles, Polygons, Polylines, Points

### [Pascal VOC](http://host.robots.ox.ac.uk/pascal/VOC/)

#### Pascal dumper description
- downloaded file: a zip archive with following structure:
  ```bash
  taskname.zip
  ├── frame_000001.xml
  ├── frame_000002.xml
  ├── frame_000003.xml
  └── ...
  ```
  Each annotation `*.xml` file has a name that corresponds to the name of the image file

>   (e.g. `frame_000001.txt` is the annotation for the `frame_000001.jpg` image).

Reviewer: ```suggestion
  (e.g. `frame_000001.xml` is the annotation for the `frame_000001.jpg` image).
```

Reviewer: Why di you ignore my comments and mark them as resolved?

  Detailed structure specification of the `*.xml` file can be found
  [here](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/devkit_doc.pdf).
- supported shapes - Rectangles
- additional comments: If you plan to use 'truncated' and 'difficult' attributes please add the corresponding
  items to the CVAT label attributes:
  `~checkbox=difficult:false ~checkbox=truncated:false`

#### Pascal loader description
-   uploaded file: a zip archive with following structure:
    ```bash
    taskname.zip
    ├── frame_000001.xml
    ├── frame_000002.xml
    ├── frame_000003.xml
    └── ...
    ```
    It should be possible to match the CVAT frame(imagename) and image filename from the annotation \*.xml

>     file(tag filename, e.g. `<filename>2008_004457.jpg</filename>`). There are 2 options:

Reviewer: Could you please correct text below?
`file(tag`

>     1. full match between image name and filename form annotation *.xml

Reviewer: form?

       file (in case of a task was created from images or archive of images).

>     1. match by frame number (if CVAT cannot match by name). File name should be in the following format `frame_%6d.jpg`.

Reviewer: Does we match by frame number now?

       It will be used when task was created from a video.

-   supported shapes: Rectangles
-   limitations: Support of Pascal VOC object detection format
-   additional comments: the CVAT task should be created with the full label set that may be in the annotation files

#### How to create a task from Pascal VOC dataset

> 1.  Download the Pascal Voc dataset

Reviewer: Please add a link

1.  Create a CVAT task with the following labels:
    ```bash
    aeroplane bicycle bird boat bottle bus car cat chair cow diningtable dog horse motorbike person pottedplant sheep sofa train tvmonitor
    ```

>     You can add `~checkbox=difficult:false ~checkbox=truncated:false` attributes if you want to use it.

Reviewer: ```suggestion
    You can add `~checkbox=difficult:false ~checkbox=truncated:false` attributes for each label if you want to use them.
```

It may take some time.

### [YOLO](https://pjreddie.com/darknet/yolo/)
#### Yolo dumper description
- downloaded file: a zip archive with following structure:
  ```bash
  taskname.zip
  ├── frame_000001.txt
  ├── frame_000002.txt
  ├── ...
  └── obj.names
  ```
  Each annotation `*.txt` file has a name that corresponds to the name of the image file
  (e.g. `frame_000001.txt` is the annotation for the `frame_000001.jpg` image).
  Short description of  `*.txt` file structure: each line describes label and bounding box
  in the following format `label_id cx cy w h`.
  `obj.names` contains the ordered list of label names.
- supported shapes - Rectangles

#### Yolo loader description
-   uploaded file: a zip archive with following structure:
    ```bash
    taskname.zip
    ├── frame_000001.txt
    ├── frame_000002.txt
    ├── frame_000003.txt
    ├── ...
    └──obj.names
    ```
    It should be possible to match the CVAT frame(imagename) and annotation filename
    There are 2 options:
    1. full match between image name and name of annotation `*.txt` file
       (in case of a task was created from images or archive of images).

>     1. match by frame number (if CVAT cannot match by name). File name should be in the following format `frame_%6d.jpg`.

Reviewer: Do we have the option now?

       It will be used when task was created from a video.

-   supported shapes: Rectangles
-   additional comments: the CVAT task should be created with the full label set that may be in the annotation files

#### How to create a task from YOLO formatted dataset (from VOC for example)
1. Follow the official [guide](https://pjreddie.com/darknet/yolo/)(see Training YOLO on VOC section)

>    and prepare the YOLO formatted annotation.

Reviewer: ```suggestion
   and prepare the YOLO formatted annotation files.
```

   ```
   Select images.zip as data. Most likely you should use `share`
   functionality because size of images.zip is more than 500Mb.
   See [Creating an annotation task](cvat/apps/documentation/user_guide.md#creating-an-annotation-task)
   guide for details.
1. Create `obj.names` with the following content:
   ```bash
   aeroplane
   bicycle
   bird
   boat
   bottle
   bus
   car
   cat
   chair
   cow
   diningtable
   dog
   horse
   motorbike
   person
   pottedplant
   sheep
   sofa
   train
   tvmonitor
   ```
1. Zip all label files together (we need to add only label files that correspond to the train subset)
   ```bash
   cat train.txt | while read p; do echo ${p%/*/*}/labels/${${p##*/}%%.*}.txt; done | zip labels.zip -j -@ obj.names
   ```
1. Click `Upload annotation` button, choose `YOLO ZIP 1.0` and select the *.zip file with labels from previous step.
   It may take some time.

### [MS COCO Object Detection](http://cocodataset.org/#format-data)
#### COCO dumper description
- downloaded file: single unpacked `json`. Detailed description of the MS COCO format can be found [here](http://cocodataset.org/#format-data)

> - supported shapes - Polygons, Rectangles (interpreted as polygons)

Reviewer: Do we interpret rectangles as polygons now?

#### COCO loader description
- uploaded file: single unpacked `*.json`.

> - supported shapes: Polygons (the `segmentation` must not be empty)

Reviewer: Is it true now?

Author: At the moment yes, I guess this description should be changed in the corresponding PR(https://github.com/opencv/cvat/pull/788)

      ```

      Select val2017.zip as data
      (See [Creating an annotation task](cvat/apps/documentation/user_guide.md#creating-an-annotation-task)
      guide for details)
1.  unpack annotations_trainval2017.zip
1.  click `Upload annotation` button,
    choose `COCO JSON 1.0` and select `instances_val2017.json.json` annotation file. It may take some time.

### [TFRecord](https://www.tensorflow.org/tutorials/load_data/tf_records)
TFRecord is a very flexible format, but we try to correspond the format that used in
[TF object detection](https://github.com/tensorflow/models/tree/master/research/object_detection)
with minimal modifications.
Used feature description:
```python

> image_feature_description = {

Reviewer: I saw a couple of fixes in the code regrading types. Could you please confirm that the definition is correct?

Author: It's up to date

to convert VOC2007 dataset to TFRecord format.
As example:
```bash
python create_pascal_tf_record.py --data_dir <path to VOCdevkit> --set train --year VOC2007 --output_path pascal.tfrecord --label_map_path label_map.pbtxt
```
1. Zip train images
   ```bash
   cat <path to VOCdevkit>/VOC2007/ImageSets/Main/train.txt | while read p; do echo <path to VOCdevkit>/VOC2007/JPEGImages/${p}.jpg  ; done | zip images.zip -j -@
   ```
1. Create a CVAT task with the following labels:
   ```bash
   aeroplane bicycle bird boat bottle bus car cat chair cow diningtable dog horse motorbike person pottedplant sheep sofa train tvmonitor
   ```
   Select images.zip as data.
   See [Creating an annotation task](cvat/apps/documentation/user_guide.md#creating-an-annotation-task)
   guide for details.
1. Zip pascal.tfrecord and label_map.pbtxt files together
   ```bash
   zip anno.zip -j <path to pascal.tfrecord> <path to label_map.pbtxt>
   ```
1. Click `Upload annotation` button, choose `TFRecord ZIP 1.0` and select the *.zip file
   with labels from previous step. It may take some time.

### PNG mask
#### Mask dumper description

> - downloaded file: a zip archive with following structure:

Reviewer: ```suggestion
- downloaded file: a zip archive with the following structure:
```
